test/XmlHttpRequestConnection.tests.js

```diff
@@ -145,6 +145,80 @@ define(function (require) {
 
         });
 
+        describe("is correctly using XmlHttpRequest and runs attached callbacks while performing:", function (){
+            beforeEach(function (){
+                mockXMLHttpRequest.prototype.send = function (body) {
+                    this.readyState = 4;
+                    this.status = 200;
+                    this.onloadstart();
+                    this.onload();
+                    this.onloadend();
+                    this.onprogress();
+                    this.ontimeout();
+                    this.onerror();
+                    this.onabort();
+                    this.upload.onloadstart();
+                    this.upload.onload();
+                    this.upload.onloadend();
+                    this.upload.onprogress();
+                    this.upload.ontimeout();
+                    this.upload.onerror();
+                    this.upload.onabort();
+                    this.onreadystatechange();
+                };
+                mockXMLHttpRequest.prototype.upload = {};
+                spyOn(mockXMLHttpRequest.prototype, 'send').andCallThrough();
+
+                window.XMLHttpRequest = (function () { return mockXMLHttpRequest; }());
+
+                connection = new XmlHttpRequestConnection();
+            });
+
+            it("execute call and invokes custom request event callbacks", function () {
```

Comment:
Well, so you are mocking up the XHR object and in fact it might look like you are testing the XHR object itself. It should be a simple mock. IMO the solution you proposed is overcomplicated and it doesn't bring much value.
I'm already testing whether provided callbacks have been called like: `expect(requestEventHandlers.onerror).toHaveBeenCalled();`. I don't see a point in reimplementing the logic of XHR object in the test case, but I might be wrong.